3party/opening_hours/opening_hours.cpp

```diff
@@ -0,0 +1,1384 @@
+{
+  return !IsEvent() && ((m_state & HaveMinutes) && !(m_state & HaveHours));
+}
+
+bool Time::IsTime() const
+{
+  return IsHoursMinutes() || IsEvent();
+}
+
+bool Time::HasValue() const
+{
+  return IsEvent() || IsTime() || IsMinutes();
+}
+
+Time Time::operator+(Time const & t)
+{
+  Time result = *this;
+  result.SetMinutes(m_duration + t.m_duration);
+  return result;
+}
+
+Time Time::operator-(Time const & t)
+{
+  Time result = *this;
+  result.SetMinutes(m_duration - t.m_duration);
+  return result;
+}
+
+Time & Time::operator-()
+{
+  m_duration = -m_duration;
+  return *this;
+}
+
+Time Time::GetEventTime() const {return {};}; // TODO(mgsergio): get real time
+
+std::ostream & operator<<(std::ostream & ost, Time::Event const event)
+{
+  switch (event)
+  {
+    case Time::Event::NotEvent:
+      ost << "NotEvent";
+      break;
+    case Time::Event::Sunrise:
+      ost << "sunrise";
+      break;
+    case Time::Event::Sunset:
+      ost << "sunset";
+      break;
+    case Time::Event::Dawn:
+      ost << "dawn";
+      break;
+    case Time::Event::Dusk:
+      ost << "dusk";
+      break;
+  }
+  return ost;
+}
+
+std::ostream & operator<<(std::ostream & ost, Time const & time)
+{
+  if (!time.HasValue())
+  {
+    ost << "hh:mm";
+    return ost;
+  }
+
+  auto const minutes = time.GetMinutesCount();
+  auto const hours = time.GetHoursCount();
+  if (time.IsEvent())
+  {
+    if (time.IsEventOffset())
+    {
+      ost << '(' << time.GetEvent();
+      if (hours < 0)
+        ost << '-';
+      else
+        ost << '+';
+      PrintPaddedNumber(ost, std::abs(hours), 2);
+      ost << ':';
+      PrintPaddedNumber(ost, std::abs(minutes), 2);
+      ost << ')';
+    }
+    else
+      ost << time.GetEvent();
+  }
+  else if (time.IsMinutes())
+    PrintPaddedNumber(ost, std::abs(minutes), 2);
+  else
+  {
+    PrintPaddedNumber(ost, std::abs(hours), 2);
+    ost << ':';
+    PrintPaddedNumber(ost, std::abs(minutes), 2);
+  }
+
+  return ost;
+}
+
+
+bool Timespan::IsEmpty() const
+{
+  return !HasStart() && !HasEnd();
+}
+
+bool Timespan::IsOpen() const
+{
+  return HasStart() && !HasEnd();
+}
+
+bool Timespan::HasStart() const
+{
+  return GetStart().HasValue();
+}
+
+bool Timespan::HasEnd() const
+{
+  return GetEnd().HasValue();
+}
+
+bool Timespan::HasPlus() const
+{
+  return m_plus;
+}
+
+bool Timespan::HasPeriod() const
+{
+  return m_period.HasValue();
+}
+
+Time const & Timespan::GetStart() const
+{
+  return m_start;
+}
+
+Time const & Timespan::GetEnd() const
+{
+  return m_end;
+}
+
+Time const & Timespan::GetPeriod() const
+{
+  return m_period;
+}
+
+void Timespan::SetStart(Time const & start)
+{
+  m_start = start;
+}
+
+void Timespan::SetEnd(Time const & end)
+{
+  m_end = end;
+}
+
+void Timespan::SetPeriod(Time const & period)
+{
+  m_period = period;
+}
+
+void Timespan::SetPlus(bool const plus)
+{
+  m_plus = plus;
+}
+
+bool Timespan::IsValid() const
+{
+  return false; // TODO(mgsergio): implement validator
```

Comment:
I'll simplify your life as it will be easy to lookup all your TODO comments and fix them.